clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -3515,4 +3668,198 @@ void handleFailure(Throwable throwable) {
+
+        Call findCoordinatorCall = getFindCoordinatorCall(context,
+            () -> KafkaAdminClient.this.getAlterConsumerGroupOffsetsCall(context, offsets, futures));
+        runnable.call(findCoordinatorCall, startFindCoordinatorMs);
+
+        return new AlterOffsetsResult(future);
+    }
+
+    private Call getAlterConsumerGroupOffsetsCall(ConsumerGroupOperationContext<Map<TopicPartition, KafkaFutureImpl<Void>>,
+            AlterOffsetsOptions> context, Map<TopicPartition,
+            OffsetAndMetadata> offsets, Map<TopicPartition, KafkaFutureImpl<Void>> futures) {
+
+        return new Call("commitOffsets", context.getDeadline(), new ConstantNodeIdProvider(context.getNode().get().id())) {
+
+            @Override
+            OffsetCommitRequest.Builder createRequest(int timeoutMs) {
+                List<OffsetCommitRequestTopic> topics = new ArrayList<>();
+                Map<String, List<OffsetCommitRequestPartition>> offsetData = new HashMap<>();
+                for (Map.Entry<TopicPartition, OffsetAndMetadata> entry : offsets.entrySet()) {
+                    String topic = entry.getKey().topic();
+                    OffsetAndMetadata oam = entry.getValue();
+                    offsetData.compute(topic, (key, value) -> {
+                        if (value == null) {
+                            value = new ArrayList<>();
+                        }
+                        OffsetCommitRequestPartition partition = new OffsetCommitRequestPartition()
+                                .setCommittedOffset(oam.offset())
+                                .setCommittedLeaderEpoch(oam.leaderEpoch().orElse(-1))
+                                .setCommittedMetadata(oam.metadata())
+                                .setCommitTimestamp(System.nanoTime())
+                                .setPartitionIndex(entry.getKey().partition());
+                        value.add(partition);
+                        return value;
+                    });
+                }
+                for (Map.Entry<String, List<OffsetCommitRequestPartition>> entry : offsetData.entrySet()) {
+                    OffsetCommitRequestTopic topic = new OffsetCommitRequestTopic()
+                            .setName(entry.getKey())
+                            .setPartitions(entry.getValue());
+                    topics.add(topic);
+                }
+                OffsetCommitRequestData data = new OffsetCommitRequestData()
+                    .setGroupId(context.getGroupId())
+                    .setTopics(topics);
+                return new OffsetCommitRequest.Builder(data);
+            }
+
+            @Override
+            void handleResponse(AbstractResponse abstractResponse) {
+                final OffsetCommitResponse response = (OffsetCommitResponse) abstractResponse;
+
+                // If coordinator changed since we fetched it, retry
+                if (context.hasCoordinatorMoved(response)) {
```

Comment:
See also `handleGroupRequestError`. If the coordinator is in the middle of being moved, we want to retry rather than fail.

Reply:
Unfortunately there's no top-level errors in OffsetCommit, so I've had to loop over the partitions again to check for group errors: https://github.com/apache/kafka/pull/7296/files#diff-6869b8fccf6b098cbcb0676e8ceb26a7R3604-R3614